src/marten/template/tag/with.cr

```diff
@@ -0,0 +1,49 @@
+require "./concerns/*"
+
+module Marten
+  module Template
+    module Tag
+      # The `with` template tag.
+      #
+      # The `with` template tag allows to assign local variables inside a block.
+      # After the block is done the assigned variables are no longer available.
+      # block:
+      #
+      # ```
+      # {% with var1=2(, var2="Hello World") %}
+      #   {{ var1 }} {{ var2 }}
+      # {% endwith %}
+      # ```
+      class With < Base
+        include CanExtractAssignments
+
+        @assignments : Array(Tuple(String, String))
+        @with_nodes : NodeSet
+
+        def initialize(parser : Parser, source : String)
+          @assignments = extract_assignments(source)
+          @with_nodes = parser.parse(up_to: {"endwith"})
+          parser.shift_token
+
+          if @assignments.size == 0
+            raise Errors::InvalidSyntax.new(
+              "Malformed with tag:at least on assignment must be present"
+            )
+          end
+        end
+
+        def render(context : Context) : String
+          context.stack do |with_context|
+            String.build do |io|
+              @assignments.each do |assignment|
+                context[assignment[0]] = FilterExpression.new(assignment.[1]).resolve(context)
```

Comment:
Ideally the `FilterExpression` should be initialized in the `#initialize` method and then resolved in the `#render` method. Indeed, `FilterExpression#new` will actually parse the raw filter expression (eg. `myvar|upcase`) and ensure that it is valid from a template syntax standpoint. If we do that in the `#render` method, this means that we will parse each filter expression every time the template is rendered!

One idea: maybe the `@assignments` instance variable could be an array of `Tuple(String, FilterExpression)` instead? This means that you would need to probably do something like that in the `#initialize` method in order to initialize the `@assignments` variable accordingly:

```crystal
def initialize(parser : Parser, source : String)
  assignments = extract_assignments(source)
  @with_nodes = parser.parse(up_to: {"endwith"})
  parser.shift_token

  if assignments.size == 0
    raise Errors::InvalidSyntax.new(
      "Malformed with tag:at least on assignment must be present"
    )
  end

  @assignemnts = assignments.map do |name, raw_filter_expression|
    {name, FilterExpression.new(raw_filter_expression)}
  end
end
```

What do  you think?

Reply:
Sounds good and reasonable 😄 thanks for having a look at it.

I will update the commit later that day